File: isf/extract_products/py/extract_products_broad.py
#!/usr/bin/env python2.7
"""Extract columns from a tab-separated summary file.

2012/10/01
Edward Liaw
"""
from __future__ import print_function
import sys
import argparse

SAVE_COLS = ('LOCUS', 'NAME')


def read_input(infile):
    """Read input.

    Args:
        infile: A filestream to read from.
    """
    pass


def mark_columns(header):
    marked = []
    spline = header.rstrip('\n').split('\t')
    for i, col in enumerate(spline):
        for save in SAVE_COLS:
            if col == save:
                marked.append(i)
    return marked


def write_output(outfile):
    """Write output.

    Args:
        outfile: A filestream to write to.
    """
    pass


def parse_arguments():
    """Handle command-line arguments.

    Returns:
        args: Arguments passed in from the command-line."""
    parser = argparse.ArgumentParser(description=__doc__,
                                     fromfile_prefix_chars='@')
    parser.add_argument('infile',
                        type=argparse.FileType('r'), nargs='?',
                        default=sys.stdin, help='input file')
    parser.add_argument('outfile',
                        type=argparse.FileType('w'), nargs='?',
                        default=sys.stdout, help='output file')
    return parser.parse_args()


def main():
    args = parse_arguments()

    # Handle I/O.
    with args.infile as infile, args.outfile as outfile:
        marked = mark_columns(next(infile))
        for line in infile:
            outstr = []
            spline = line.rsplit('\n').split('\t')
            for mark in marked:
                outstr.append(spline[mark])
            print('\t'.join(outstr), file=outfile)


if __name__ == "__main__":
    main()

File: isf/extract_products/pl/extract_products_broad.pl
#!/usr/bin/env perl
# Extract genes and product names from genome summary file.
# 2012-09-10
# Edward Liaw
use strict;
use warnings;

# Command line args.
$#ARGV == 0 or die "Too few arguments";
open INFILE, '<', $ARGV[0] or die "Can't open $ARGV[0] for reading!";

# Find the columns we want to save.
chomp(my $header = <INFILE>);
my @headings = split(/\t/, $header);
my $tab = 0;
my $locus_index = 0;
my $name_index = 0;
foreach my $heading (@headings) {
	if ($heading eq "LOCUS") {
		$locus_index = $tab;
	}
	elsif ($heading eq "NAME") {
		$name_index = $tab;
	}
	$tab++;
}

while (my $line = <INFILE>) {
	chomp $line;
	my @spline = split(/\t/, $line);
	print "$spline[$locus_index]\t$spline[$name_index]\n";
}

close INFILE;
